plutus-core/plutus-core/src/PlutusCore/Builtin/KnownTypeAst.hs

```diff
@@ -0,0 +1,269 @@
+Plutus and they don't exist in the type language of Plutus.
+
+These @*Rep@ data families like 'TyVarRep', 'TyAppRep' etc all require the Rep context and preserve
+it, since they're only for representing Plutus types for type checking purposes.
+
+We call a thing in a Rep or 'Type' context a 'RepHole' or 'TypeHole' respectively. The reason for
+the name is that the inference machinery looks at the thing and tries to instantiate it, like fill
+a hole.
+
+We could also have a third type of hole/context, Name, because binders bind names rather than
+variables and so it makes sense to infer names sometimes, like for 'TyForallRep' for example.
+We don't do that currently, because we don't have such builtins anyway.
+
+And there could be even fancier kinds of holes like "infer anything" for cases where the hole
+is determined by some other part of the signature. We don't have that either, for the same reason.
+
+For the user defining a builtin this all is pretty much invisible.
+-}
+
+-- See Note [Rep vs Type context].
+-- | The kind of holes.
+data Hole
+
+-- See Note [Rep vs Type context].
+-- | A hole in the Rep context.
+type RepHole :: forall a hole. a -> hole
+data family RepHole x
+
+-- See Note [Rep vs Type context].
+-- | A hole in the Type context.
+type TypeHole :: forall hole. GHC.Type -> hole
+data family TypeHole a
+
+-- | For annotating an uninstantiated built-in type, so that it gets handled by the right instance
+-- or type family.
+type BuiltinHead :: forall k. k -> k
+data family BuiltinHead f
+
+-- | Take an iterated application of a built-in type and elaborate every function application
+-- inside of it to 'TypeAppRep', plus annotate the head with 'BuiltinHead'.
+-- The idea is that we don't need to process built-in types manually if we simply add some
+-- annotations for instance resolution to look for. Think what we'd have to do manually for, say,
+-- 'ToHoles': traverse the spine of the application and collect all the holes into a list, which is
+-- troubling, because type applications are left-nested and lists are right-nested, so we'd have to
+-- use accumulators or an explicit 'Reverse' type family. And then we also have 'KnownTypeAst' and
+-- 'ToBinds', so handling built-in types in a special way for each of those would be a hassle,
+-- especially given the fact that type-level Haskell is not exactly good at computing things.
+-- With the 'ElaborateBuiltin' approach we get 'KnownTypeAst', 'ToHoles' and 'ToBinds' for free.
+type ElaborateBuiltin :: forall k. k -> k
+type family ElaborateBuiltin a where
+    ElaborateBuiltin (f x) = ElaborateBuiltin f `TyAppRep` x
+    ElaborateBuiltin f     = BuiltinHead f
+
+-- | A constraint for \"@a@ is a 'KnownTypeAst' by means of being included in @uni@\".
+type KnownBuiltinTypeAst uni a = KnownTypeAst uni (ElaborateBuiltin a)
+
+type KnownTypeAst :: forall a. (GHC.Type -> GHC.Type) -> a -> GHC.Constraint
+class KnownTypeAst uni x where
```

Comment:
Just noticed that this class is unusual in missing some haddock!

Reply:
I'll fix that separately.